File: Assignment 1/ques2.cpp
#include <iostream>
#include <vector>
#include <math.h>
#include <limits.h>

using namespace std;

long long int inversions;

void merge_inversions(vector<long long int> &inputs, long long int p, long long int q, long long int r)
{
    long long int n1 = q - p;
    long long int n2 = r - q;

    vector<long long int> left, right;
    for (long long int i = 0; i < n1; i++)
    {
        left.push_back(inputs[p + i]);
    }
    for (long long int i = 0; i < n2; i++)
    {
        right.push_back(inputs[q + i]);
    }

    left.push_back(INT_MAX);
    right.push_back(INT_MAX);
    long long int i = 0;
    long long int j = 0;
    for (long long int k = p; k < r; k++)
    {
        if (left[i] <= right[j])
        {
            inputs[k] = left[i];
            i++;
        }
        else
        {
            inputs[k] = right[j];
            j++;
            inversions += n1 - i;
        }
    }
}

void count_inversions(vector<long long int> &inputs, long long int p, long long int r)
{
    if (p >= r - 1)
        return;
    long long int q = (p + r) / 2;
    count_inversions(inputs, p, q);
    count_inversions(inputs, q, r);
    merge_inversions(inputs, p, q, r);
}

int main()
{
    long long int n;
    cin >> n;

    vector<long long int> inputs;

    long long int temp;
    for (long long int idx = 0; idx < n; idx++)
    {
        cin >> temp;

        inputs.push_back(temp);
    }
    count_inversions(inputs, 0, (long long int)inputs.size());
    cout << inversions << endl;
    return 0;
}

File: Assignment 1/ques2.py
import sys
sys.setrecursionlimit(10**6)

def merge_inversions(left, right):
    result = []
    i= 0
    j =0
    inversions = 0

    ll = len(left)
    lr = len(right)
    
    while i < ll and j < lr:
        
        if left[i] <= right[j]:
            result.append(left[i])
            i = i + 1
        
        else:
            result.append(right[j])
            j = j + 1
            inversions = inversions + ll-i
    
    result = result + left[i:]
    result = result + right[j:]
    
    return result,inversions

def count_inversions(array):
    if len(array) <= 1:
        return array, 0
    middle = len(array) // 2
    
    left,inv_left = count_inversions(array[:middle])
    right,inv_right = count_inversions(array[middle:])
    merged, inversions = merge_inversions(left,right)
    
    inversions = inversions + inv_left + inv_right
    return merged, inversions

def inv(array):
    _, inversions = count_inversions(array)
    return inversions

if __name__ == "__main__":
    n = int(input())
    array = [int(x) for x in input().strip().split(" ")]
    inversions = inv(array)
    print(inversions)